x-pack/plugins/alerting/README.md

```diff
@@ -0,0 +1,297 @@
+		const cpuUsageByServer = await getCpuUsageByServer();
+
+		for (const { server, cpuUsage: currentCpuUsage } of cpuUsageByServer) {
+			// Only fire if CPU usage is greater than threshold
+			if (currentCpuUsage > threshold) {
+				// The first argument is a unique identifier the alert instance is about. In this scenario
+				// the provided server will be used. Also, this id will be used to make `getState()` return
+				// previous state, if any, on matching identifiers.
+				const alertInstance = services.alertInstanceFactory(server);
+
+				// State from last execution. This will exist if an alert instance was created and fired
+				// in the previous execution
+				const { cpuUsage: previousCpuUsage } = alertInstance.getState();
+
+				// Replace state entirely with new values
+				alertInstance.replaceState({
+					cpuUsage: currentCpuUsage,
+				});
+
+				// 'default' refers to a group of actions to fire, see 'actions' in create alert section
+				alertInstance.fire('default', {
+					server,
+					hasCpuUsageIncreased: currentCpuUsage > previousCpuUsage,
+				});
+			}
+		}
+
+		// Single object containing state that isn't specific to a server, this will become available
+		// within the `state` function parameter at the next execution
+		return {
+			lastChecked: new Date(),
+		};
+	},
+});
+```
+
+## RESTful API
+
+Using an alert type requires you to create an alert that will contain parameters and actions for a given alert type. See below for CRUD operations using the API.
+
+#### `POST /api/alert`: Create alert
+
+Payload:
+
+|Property|Description|Type|
+|---|---|---|
+|alertTypeId|The id value of the alert type you want to call when the alert is scheduled to execute.|string|
+|interval|The interval in milliseconds the alert should execute.|number|
+|alertTypeParams|The parameters to pass in to the alert type executor `params` value. This will also validate against the alert type params validator if defined.|object|
+|actions|Array of the following:<br> - `group` (string): We support grouping actions in the scenario of escalations or different types of alert instances. If you don't need this, feel free to use `default` as a value.<br>- `id` (string): The id of the action saved object to fire.<br>- `params` (object): The map to the `params` the action type will receive. In order to help apply context to strings, we handle them as mustache templates and pass in a default set of context. (see templating actions).|array|
+
+#### `DELETE /api/alert/{id}`: Delete alert
+
+Params:
+
+|Property|Description|Type|
+|---|---|---|
+|id|The id of the alert you're trying to delete.|string|
+
+#### `GET /api/alert/_find`: Find alerts
+
+Params:
+
+See the saved objects API documentation for find. All the properties are the same except you cannot pass in `type`.
+
+#### `GET /api/alert/{id}`: Get alert
+
+Params:
+
+|Property|Description|Type|
+|---|---|---|
+|id|The id of the alert you're trying to get.|string|
```

Comment:
Some information on what gets returned would be useful - I think we're just returning the alert configuration (i.e. what is in the saved object).  Some people might be expecting instance data like  alert state being returned here (say I want to get my alert and show its state in UI).  But sounds like we'd need to get that via the alertInstanceFactory since it is part of the instance.

Reply:
Yeah, I can see that. We can add new APIs or modify them as we build the alerting app to display valuable information. We may have to move things around or develop something new to make the data accessible.